### PR TITLE
feat(llm): convert function call request for non-funcall OSS model

### PR DESCRIPTION
**Pull request by xingyaoww**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This PR adds a general utility that automatically converts **function-calling LLM requests** to **non-function-calling LLM requests** under the hood.

- This get rid of the need to maintain two set of prompts for (1) function calling and (2) non-function calling models, and will greatly reduce our maintenance burden. 
- **Function-calling is now by default ON**: Going forward, we only need to iterate on "function calling mode" only, and the "non-function calling backward-compatibility" will automatically happen under the hood.
- We now curate a list of "supported function calling model" in `llm.py` based on our evaluation result below:

Evaluation results so far:

- Claude is still the default to-go model :)
- Llama 3.1 70B & Gemini-002-pro barely worked in function calling mode, but works much better with non-function calling
- For OSS: Llama, Qwen, Deepseek are all good options (though with low resolve rate)

image

---
**Link of any specific issues this addresses**

Should fix https://github.com/All-Hands-AI/OpenHands/issues/4865

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:628201b-nikolaik   --name openhands-app-628201b   docker.all-hands.dev/all-hands-ai/openhands:628201b
```

### CODE REVIEW

File: openhands/llm/fn_call_converter.py [hunk 1]
```diff
@@ -0,0 +1,795 @@
...
+multiple lines
+</parameter>
+</function>
+
+<IMPORTANT>
+Reminder:
+- Function calls MUST follow the specified format, start with <function= and end with </function>
+- Required parameters MUST be specified
+- Only call one function at a time
+- You may provide optional reasoning for your function call in natural language BEFORE the function call, but NOT after.
```
Comment (enyst) [inline openhands/llm/fn_call_converter.py:43]: This sounds reasonable, but we could also be lenient with it? Like we could discard whatever follows /function.

File: openhands/llm/fn_call_converter.py [hunk 2]
```diff
@@ -0,0 +1,795 @@
...
+- If there is no function call available, answer the question like normal with your current knowledge and do not tell the user about function calls
+"""
+
+STOP_WORDS = ['</function']
```
Comment (enyst) [inline openhands/llm/fn_call_converter.py:47]: Aha, already we discard here

File: openhands/llm/fn_call_converter.py (hunk 1 above)
Comment (xingyaoww) [reply openhands/llm/fn_call_converter.py:43]: yeah - we set stop word to `</function`

**Comment by enyst** [comment]
This is a great idea and a great PR. We should be doing this, it will be better to clear out that FC/non-FC code...

However, I do have some (unbaked) thoughts. For the sake of clarity I'll express them a bit rough, even though it's hard to be sure of this kind of stuff.

I think those results are interesting because they ... don't match expectations. My intuitions were that 
1) most, not all, but most LLMs that natively support fc would work the same or a bit better
2) gpt-4o fc results would be below, but visibly in the "same class"
3) gemini (!) fc results would be below, but visibly in roughly the "same class" (1% !  👀 🫨)
4) fc or non-fc, _some_ OSS results would be sort of close

I think a possible explanation of these results, in part, is that we are now doing a fairly extreme version of optimizing prompts for Claude. We're not just "not helping" these other LLMs, we might be prompting in some ways that are "bad" for them.

Just for illustration, a few examples of what I mean: 
- Deepseek with browsing agent [can respond correctly, but fail the task](https://github.com/All-Hands-AI/OpenHands/issues/4355#issuecomment-2408787650): it responded with a message sent with python variables, when our framework expected just a text. Note: it did follow instructions. We just didn't see Claude responding this way to these instructions so our code assumed it won't happen.
- Gemini falls into [this stuck scenario](https://github.com/All-Hands-AI/OpenHands/blob/27cd507cd21ae8349df7080a0be26d7ea2ae8142/openhands/controller/stuck.py#L231). I find this interesting because I've never seen CodeAct hit that before. That code was literally, IMO, dead code, for months, it was there since monologue agent's times, and I honestly thought to remove it (or keep it for the amazingly precise reason "in theory it's not reaaaaally impossible" 😅).
IMHO, what this says is that our CodeAct prompts are enough for GPT-4o to not hit that. But they're not appropriate enough for Gemini.

I feel like we may need to consider prompting better and test/respond better for like 3-4 LLMs, of which at least one OSS...

But I don't know if/how we may want to square this circle, to both:
- try to give at least 3-4 LLMs other than Claude a fair chance to show us what they can do, 
- and in the same time ... I do agree that maintaining 2 sets of prompts is not great... let alone 3-4.

Other note:
- have we looked at Nous Research Llamas? I would hope we can try it sometime. And/or Llama 3.2 ?

**Review by rbren — APPROVED** [review]

**Comment by manarbenkraouda** [comment]
Hey @xingyaoww ! Have you seen this? https://x.com/ollama/status/1861442333385892026
I think it may be relevant to this PR!

**Comment by enyst** [comment]
> > gemini (!) fc results would be below, but visibly in roughly the "same class" (1% ! 👀 🫨)
> 
> For Gemini, the bad function calling result is more likely a bug / artifact
> 
> It is able to call tools correctly early in the interaction 
> But it starts to add a weird "fields" to the tool call later in the same trajectory 😓 

I looked into the Gemini native API and [its docs](https://github.com/google-gemini/cookbook/blob/main/quickstarts/Function_calling.ipynb) mention 'fields'. I guess the model must have been trained with `fields` for some reason 🫣

**Comment by BradKML** [comment]
So what can be done to make FC-free tooling for Open Weight models (Llama, DeepSeek, Qwen) to be as good as the big APIs with function calls?

**Comment by BradKML** [comment]
Also a bit of self-note regarding function call benchmarks:
- Berkeley classic https://gorilla.cs.berkeley.edu/leaderboard.html
- https://github.com/ComposioHQ/Composio-Function-Calling-Benchmark
- Nested function calls https://github.com/IBM/NESTFUL
- Nexus and NFCL? https://www.databricks.com/blog/unpacking-function-calling-eval

**Pull request merged**